flink-java/src/main/java/org/apache/flink/api/java/typeutils/EitherTypeInfo.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.java.typeutils;
+
+import org.apache.flink.api.common.ExecutionConfig;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.java.typeutils.runtime.EitherSerializer;
+
+/**
+ * A {@link TypeInformation} for the {@link Either} type of the Java API.
+ *
+ * @param <L> the Left value type
+ * @param <R> the Right value type
+ */
+public class EitherTypeInfo<L, R> extends TypeInformation<Either<L, R>> {
+
+	private static final long serialVersionUID = 1L;
+
+	private final Class<Either<L, R>> typeClass;
+
+	private final TypeInformation<L> leftType;
+
+	private final TypeInformation<R> rightType;
+
+	@SuppressWarnings({ "unchecked", "rawtypes" })
+	public EitherTypeInfo(Class<Either> typeClass,
+			TypeInformation<L> leftType,TypeInformation<R> rightType) {
+		this.typeClass = (Class<Either<L, R>>) (Class<?>) typeClass;
+		this.leftType = leftType;
+		this.rightType = rightType;
+	}
+
+	@Override
+	public boolean isBasicType() {
+		return false;
+	}
+
+	@Override
+	public boolean isTupleType() {
+		return false;
+	}
+
+	@Override
+	public int getArity() {
+		return 1;
+	}
+
+	@Override
+	public int getTotalFields() {
+		return 1;
+	}
+
+	@Override
+	public Class<Either<L, R>> getTypeClass() {
+		return this.typeClass;
```

Comment:
I would change this to `return (Class<Either<L, R>>) (Class<?>) Either.class`. You can then get rid of the field `this.typeClass`.